src/main/java/com/fasterxml/jackson/databind/type/TypeBindings.java

```diff
@@ -309,11 +317,17 @@ public boolean hasUnbound(String name) {
      * Factory method that will create an object that can be used as a key for
      * caching purposes by {@link TypeFactory}
      *
+     * @return An object which can be used as a key in TypeFactory, or {@code null} if no key can be created.
+     *
      * @since 2.8
      */
     public Object asKey(Class<?> rawBase) {
         // safe to pass _types array without copy since it is not exposed via
         // any access, nor modified by this class
+        if (containsPlaceholders()) {
```

Comment:
Is there any chance that this PR could be split in 2? The hashCode fixes seem useful but are probably more risky than this specific fix ('containsPlaceholders()'). The hashCode PR could be merged but maybe not given as a high a priority for release.

@cowtowncoder these changes seem reasonable generally but I think we reduce the risks in the v2.15.0 release if we prioritise smaller changes

Reply:
@pjfanning That's a terribly difficult call to make because both pieces directly contribute to hash collisions in the cache which result in lock contention within ConcurrentHashMap buckets.

Reply:
I can't see how leaving items out of the cache and changing hash code implementations have to be shoe horned into 1 PR. Even if the reasons for the changes overlap.

Reply:
Ok I think things make sense now; I wish smaller changes suffice but understanding the challenge of collisions makes me think what remains is actually necessary. And while timing is not great we do get one more RC; and I think functionality is reasonably well covered due to how widely `JavaType`s are used.
So I think I'll want to go with it, get it merged, push rc3.

I'm more hesitant wrt 2.14 due to additional type introduced (f.ex).